include/partition_bins.h

```diff
@@ -0,0 +1,215 @@
+    delete[] recvPartitions;
+  }
+
+  // double accum_time, accum_init;
+  // int accum_len;
+  // int accum_res;
+
+  void cleanDebugStorage() {
+    // accum_init = 0.0;
+    // accum_time = 0.0;
+    // accum_len = 0;
+    // accum_res = 0;
+  }
+
+  void printDebugStorage() {
+    // std::cout << "(" << mpi_rank << ") " << accum_init << " " << accum_time << " " << accum_len << " " << accum_res <<std::endl;
+  }
+
+  void SearchInRadius(TObjectType const& ThePoint, double Radius, std::vector<ResultType>& rResults) {
+    InternalPointType min_point;
+    std::array<std::size_t, Dimension> length;
+
+    // double t0, t1;
+
+    // t0 = GetCurrentTime();
+    for (int i = 0; i < Dimension; i++) {
+      min_point[i] = ThePoint[i] - Radius;
+      length[i] = mCells.CalculatePosition(ThePoint[i] + Radius, i) - mCells.CalculatePosition(ThePoint[i] - Radius, i) + 1;
+    }
+    auto min_cell = mCells.CalculateCellIndex(min_point);
+
+    std::vector<int> usedMask(mpi_size, 0);
+    // t1 = GetCurrentTime();
+
+    // accum_init += (t1 - t0);
+
+    // t0 = GetCurrentTime();
+    for (std::size_t i_z = 0; i_z < length[2]; i_z++) {
+      auto y_position = min_cell + i_z * mCells.GetNumberOfCells(0) * mCells.GetNumberOfCells(1);
+      for (std::size_t i_y = 0; i_y < length[1]; i_y++) {
+        std::size_t offset = mCells.GetCellBeginIndex(y_position);
+        const std::size_t end_offset = mCells.GetCellBeginIndex(y_position + length[0]);
+        int * p_part = mpPartitions + mCells.GetCellBeginIndex(y_position);
+        for (; offset <end_offset; offset++) {
+          if(!usedMask[*p_part] && *p_part != -1)
+          { 
+            usedMask[*p_part] = 1;
+            rResults.push_back(*p_part);
+          }
+          p_part++;
+        }
+        y_position += mCells.GetNumberOfCells(0);
+      }
+    }
+    // t1 = GetCurrentTime();
+
+    // accum_time += (t1 - t0);
+    // accum_len += (length[2] * length[1]);
+    // accum_res += rResults.size();
+  }
+
+private:
+  std::size_t mNumberOfPoints;
+  PartitionBinsCellsContainer mCells;
+  int * mpPartitions;
+
+  template<typename TIteratorType>
+  void AssignPartitionsToCells(TIteratorType const& PointsBegin, TIteratorType const& PointsEnd, int * recvOffsets, std::size_t numberOfCells) {
+    for (auto i_point = PointsBegin; i_point != PointsEnd; i_point++) {
+      auto index = mCells.CalculateCellIndex(*i_point);
+      std::size_t offset = mCells.GetCellBeginIndex(index);
+
+      for(int p = 0; p < mpi_size; p++) {
+        if((recvOffsets[numberOfCells * p + index + 1] - recvOffsets[numberOfCells * p + index]) == 1) {
+          mpPartitions[offset++] = p;
+        }
+      }
+    }
+  }
+
+  double Distance2(TObjectType const& FirstPoint, TObjectType const& SecondPoint) {
+    double result = double();
+    for (int i = 0; i < Dimension; i++) {
+      auto distance_i = FirstPoint[i] - SecondPoint[i];
+      result += distance_i * distance_i;
+    }
+    return result;
+  }
+};
```

Comment:
we should find a way to reuse the code between this class and points bins and objects bins.

Reply:
I will open a issue for that once merged.